component/attr_cache/attr_cache.go

```diff
@@ -403,26 +441,84 @@ func (ac *AttrCache) StreamDir(options internal.StreamDirOptions) ([]*internal.O
-		ac.cacheAttributes(pathList)
-
-		// merge missing directory cache into the last page of results
-		if ac.cacheDirs && token == "" {
-			var numAdded int // prevent shadowing pathList in following line
-			pathList, numAdded = ac.addDirsNotInCloudToListing(options.Name, pathList)
-			log.Trace("AttrCache::StreamDir : %s +%d from cache = %d",
-				options.Name, numAdded, len(pathList))
+		// cache returned list
+		ac.cacheAttributes(pathList, options.Name, token)
+		//
+		if ac.cacheDirs {
+			// remember that this directory is in cloud storage
+			if len(pathList) > 0 {
+				ac.markAncestorsInCloud(options.Name, time.Now())
+			}
+			// merge missing directory cache into the last page of results
+			if ac.cacheDirs && token == "" {
+				var numAdded int // prevent shadowing pathList in following line
+				pathList, numAdded = ac.addDirsNotInCloudToListing(options.Name, pathList)
+				log.Info("AttrCache::StreamDir : %s +%d from cache = %d",
+					options.Name, numAdded, len(pathList))
+			}
 		}
 	}
+	// add cached items in
+	if len(cachedPathList) > 0 {
+		log.Info("AttrCache::StreamDir : %s merging in %d list cache entries...", options.Name, len(cachedPathList))
+		pathList = append(pathList, cachedPathList...)
+	}
+	// values should be returned in ascending order by key, without duplicates
+	// sort
+	slices.SortFunc[[]*internal.ObjAttr, *internal.ObjAttr](pathList, func(a, b *internal.ObjAttr) int {
+		return strings.Compare(a.Path, b.Path)
+	})
+	// remove duplicates
+	pathList = slices.CompactFunc[[]*internal.ObjAttr, *internal.ObjAttr](pathList, func(a, b *internal.ObjAttr) bool {
+		return a.Path == b.Path
+	})
 
+	log.Trace("AttrCache::StreamDir : %s returning %d entries", options.Name, len(pathList))
 	return pathList, token, err
 }
 
+// Return directory listing from cache
+// Any request other than a request for the next page will return all children,
+// and the token for the next page (if there is one).
+// If page requests are repeated or backtrack, this may cause unexpected OS behavior.
+func (ac *AttrCache) fetchCachedDirList(path string, token string) ([]*internal.ObjAttr, string, error) {
+	log.Trace("AttrCache::fetchCachedDirList : %s token=\"%s\"", path, token)
+
+	var pathList []*internal.ObjAttr
+	listDirCache, getErr := ac.cacheMap.get(path)
+	if getErr != nil {
+		log.Debug("AttrCache::fetchCachedDirList : %s directory not found in cache", path)
+		return pathList, "", fmt.Errorf("%s directory not found in cache", path)
+	}
+	log.Debug("AttrCache::fetchCachedDirList : %s listing token=\"%s\"", path, listDirCache.listToken)
+	// check timeout
+	if time.Since(listDirCache.listedAt).Seconds() >= float64(ac.cacheTimeout) {
+		log.Debug("AttrCache::fetchCachedDirList : %s listing cache expired", path)
+		return pathList, "", fmt.Errorf("%s directory listing expired", path)
+	}
+	// don't provide cached data when new (uncached) data is being requested
+	if token != "" && token == listDirCache.listToken {
```

Comment:
It's a bit confusing, honestly.
Yes, I cache the returned items and record the token that came back. But since the token that came back is meant to be used to get more entries _**from the cloud**_, when a subsequent request has a matching token, I **do not** serve from cache but route that request to the cloud.